File: validate-binary-search-tree.cpp
//Runtime: 9 ms
//Learned from "discuss" section.
class Solution {
public:

    bool validate(TreeNode* root, TreeNode* minn, TreeNode* maxn)
    {
        if(!root) return 1;

        if((minn && root->val <= minn->val) ||(maxn && root->val >= maxn->val))
            return 0;

        return (validate(root->left, minn, root) && validate(root->right, root, maxn));
    }
    bool isValidBST(TreeNode* root) {

        return validate(root, NULL, NULL);
    }
};

File: validate-binary-search-tree.scala
//Runtime: 516 ms
object Solution {
    def isValidBST(root: TreeNode): Boolean = {
        def solve(root:TreeNode, a:TreeNode, b:TreeNode): Boolean = root match {
            case null => true
            case t:TreeNode =>
                if((a != null && root.value <= a.value) || (b != null && root.value >= b.value)) false
                else solve(root.left, a, root) && solve(root.right, root, b)
        }
        solve(root, null, null)
    }
}